### PR TITLE
RUBY-324: Handle frozen collections correctly

### PR DESCRIPTION
**Pull request by mnin**
https://datastax-oss.atlassian.net/browse/RUBY-324

### CODE REVIEW

**Comment by mnin** [comment]
I cannot access the `datastax.lan` URL from the job-creator check. How to solve this problem?

**Comment by stamhankar999** [comment]
I'm afraid that server isn't accessible to non-DataStax folks. However, I checked it out, and the failure is an infrastructure issue that I'm trying to fix now. However, you do need to sign the CLA before we can proceed with the PR.

**Comment by stamhankar999** [comment]
Ok, I believe I've fixed the infrastructure. Please sign the CLA and make a no-op commit to trigger a new CI build.

**Comment by mnin** [comment]
@stamhankar999 thanks for your answer, I signed the CLA already

**Comment by mnin** [comment]
@stamhankar999 is the infrastructure problem solved?

File: spec/cassandra/cluster/schema/fetchers/2.2.1-data.json [hunk 1]
```diff
@@ -3035,6 +3035,17 @@
...
+    {
+      "keyspace_name": "simplex",
+      "columnfamily_name": "t3",
+      "column_name": "f3",
+      "component_index": 0,
+      "index_name": null,
+      "index_options": "null",
+      "index_type": null,
+      "type": "regular",
+      "validator": "org.apache.cassandra.db.marshal.ListType(org.apache.cassandra.db.marshal.FrozenType(org.apache.cassandra.db.marshal.Int32Type))"
```
Comment (stamhankar999) [inline spec/cassandra/cluster/schema/fetchers/2.2.1-data.json]: This looks backwards. Shouldn't it be `org.apache.cassandra.db.marshal.FrozenType(org.apache.cassandra.db.marshal.ListType(org.apache.cassandra.db.marshal.Int32Type))`?

File: spec/cassandra/cluster/schema/fetchers/2.2.1-schema.cql [hunk 2]
```diff
@@ -708,7 +708,8 @@ WITH CLUSTERING ORDER BY ("f2" DESC)
 
 CREATE TABLE simplex."t3" (
   "f1" int PRIMARY KEY,
-  "f2" int
+  "f2" int,
+  "f3" list<frozen<int>>
```
Comment (stamhankar999) [inline spec/cassandra/cluster/schema/fetchers/2.2.1-schema.cql]: And thus, this should be `frozen<list<int>>`.

File: spec/cassandra/cluster/schema/fetchers/3.0.0-data.json [hunk 3]
```diff
@@ -306,6 +306,16 @@
...
+    },
+    {
+      "keyspace_name": "simplex",
+      "table_name": "t4",
+      "column_name": "f4",
+      "clustering_order": "none",
+      "column_name_bytes": "f3",
+      "kind": "regular",
+      "position": 0,
+      "type": "list<frozen<int>>"
```
Comment (stamhankar999) [inline spec/cassandra/cluster/schema/fetchers/3.0.0-data.json]: This is similarly invalid; should be `frozen<list<int>>`.

File: spec/cassandra/cluster/schema/fetchers/3.0.0-schema.cql [hunk 4]
```diff
@@ -72,6 +72,7 @@ CREATE TABLE simplex."t4" (
   "f1" int,
   "f2" int,
   "f3" int,
+  "f4" list<frozen<int>>,
```
Comment (stamhankar999) [inline spec/cassandra/cluster/schema/fetchers/3.0.0-schema.cql]: Same here.

**Comment by stamhankar999** [comment]
Please merge down from master; there is a change in build.yaml that you need in order for the Jenkins build to run. Other than the test fixes and that, I think this is good to go. Thanks for the contribution!

File: spec/cassandra/cluster/schema/fetchers/2.2.1-data.json (hunk 1 above)
Comment (mnin) [reply spec/cassandra/cluster/schema/fetchers/2.2.1-data.json]: Nope, it should be e.g. exactly this: `list<frozen <list<int>>>`.

See https://docs.datastax.com/en/cql/3.3/cql/cql_reference/collection_type_r.html

**Comment by mnin** [comment]
Here's a better example of what we do:

Setup:
```
CREATE KEYSPACE k1 WITH REPLICATION = { 'class' : 'SimpleStrategy', 'replication_factor' : 1 };

CREATE TABLE k1.t1 (
  id bigint PRIMARY KEY,
  list list<frozen<tuple<bigint, text>>>
);

INSERT INTO k1.t1 (id, list) values(1, [(1, 'foo'), (2, 'foo')]);

CREATE TABLE k1.t2 (
  id bigint PRIMARY KEY,
  list frozen<list<tuple<bigint, text>>>
);

INSERT INTO k1.t2 (id, list) values(1, [(1, 'foo'), (2, 'foo')]);
```

```
cqlsh> UPDATE k1.t1 SET list = [(1, 'bar')] + list WHERE id = 1;

cqlsh> UPDATE k1.t2 SET list = [(1, 'bar')] + list WHERE id = 1;
InvalidRequest: Error from server: code=2200 [Invalid query] message="Invalid operation (list = [(1, 'bar')] - list) for frozen list column list"
```

File: spec/cassandra/cluster/schema/fetchers/2.2.1-data.json (hunk 1 above)
Comment (stamhankar999) [reply spec/cassandra/cluster/schema/fetchers/2.2.1-data.json]: But that leads to this validator in C* (when I try with C* 2.2.9):
`org.apache.cassandra.db.marshal.ListType(org.apache.cassandra.db.marshal.FrozenType(org.apache.cassandra.db.marshal.ListType(org.apache.cassandra.db.marshal.Int32Type)))`

The validator value you have above puts FrozenType around Int32Type, which I think is illegal. You can't freeze a primitive, only a collection.

File: spec/cassandra/cluster/schema/fetchers/2.2.1-schema.cql (hunk 2 above)
Comment (stamhankar999) [reply spec/cassandra/cluster/schema/fetchers/2.2.1-schema.cql]: Analogously, `list<frozen<int>>` is illegal / not possible in reality.

File: spec/cassandra/cluster/schema/fetchers/2.2.1-data.json (hunk 1 above)
Comment (stamhankar999) [reply spec/cassandra/cluster/schema/fetchers/2.2.1-data.json]: Just confirmed that 2.2.1 also produces the same validator string.

**Comment by mnin** [comment]
@stamhankar999 fixed in ff8966f

**Pull request merged**